streams-contrib/streams-persist-cassandra/src/main/java/org/apache/streams/cassandra/CassandraPersistReader.java

```diff
@@ -0,0 +1,316 @@
+      Map<String, Object> replication = new HashMap<>();
+      replication.put("class", "SimpleStrategy");
+      replication.put("replication_factor", 1);
+
+      String createKeyspaceStmt = SchemaBuilder.createKeyspace(keyspace).with()
+          .replication(replication).getQueryString();
+      cluster.connect().execute(createKeyspaceStmt);
+    }
+
+    session = cluster.connect(keyspace);
+
+    KeyspaceMetadata ks = metadata.getKeyspace(keyspace);
+    TableMetadata tableMetadata = ks.getTable(table);
+
+    if (Objects.isNull(tableMetadata)) {
+      LOGGER.info("Table {} does not exist in Keyspace {}. Creating Table", table, keyspace);
+      String createTableStmt = SchemaBuilder.createTable(table)
+                                .addPartitionKey(config.getPartitionKeyColumn(), DataType.varchar())
+                                .addColumn(config.getColumn(), DataType.blob()).getQueryString();
+
+      session.execute(createTableStmt);
+    }
+  }
+
+  @Override
+  public StreamsResultSet readAll() {
+    ResultSet rs = session.execute(getSelectStatement());
+    Iterator<Row> rowsIterator = rs.iterator();
+
+    while (rowsIterator.hasNext()) {
+      Row row = rowsIterator.next();
+      StreamsDatum datum = prepareDatum(row);
+      write(datum);
+    }
+
+    return readCurrent();
+  }
+
+  @Override
+  public void startStream() {
+    LOGGER.debug("startStream");
+    CassandraPersistReaderTask readerTask = new CassandraPersistReaderTask(this);
+
+    CompletableFuture.runAsync(readerTask, executor);
+
+    try {
+      if (readerTaskFuture.get()) {
+        executor.shutdown();
+      }
+    } catch (InterruptedException ex) {
+      LOGGER.trace("Interrupt", ex);
+    } catch (ExecutionException ex) {
+      LOGGER.trace("Execution exception", ex);
+    }
+  }
+
+  @Override
+  public StreamsResultSet readCurrent() {
+
+    StreamsResultSet current;
+
+    try {
+      lock.writeLock().lock();
+      current = new StreamsResultSet(persistQueue);
+      current.setCounter(new DatumStatusCounter());
+      persistQueue = constructQueue();
+    } finally {
+      lock.writeLock().unlock();
+    }
+
+    return current;
+  }
+
+  protected void write(StreamsDatum entry) {
+    boolean success;
+    do {
+      try {
+        lock.readLock().lock();
+        success = persistQueue.offer(entry);
+        Thread.yield();
+      } finally {
+        lock.readLock().unlock();
+      }
+    }
+    while (!success);
+  }
+
+  @Override
+  public StreamsResultSet readNew(BigInteger sequence) {
+    return null;
+  }
+
+  @Override
+  public StreamsResultSet readRange(DateTime start, DateTime end) {
+    return null;
+  }
+
+  @Override
+  public boolean isRunning() {
+    return !executor.isTerminated() || !executor.isShutdown();
+  }
+
+  private Queue<StreamsDatum> constructQueue() {
+    return Queues.synchronizedQueue(new LinkedBlockingQueue<StreamsDatum>(10000));
```

Comment:
Question: does the Queue need to be synchronized if not using any of the bulk Queue API ?